numba/targets/arraymath.py

```diff
@@ -914,6 +914,111 @@ def nanmedian_impl(arry):
 #----------------------------------------------------------------------------
 # Element-wise computations
 
+@register_jitable
+def _fill_diagonal_params(a, wrap):
+    if a.ndim == 2:
+        m = a.shape[0]
+        n = a.shape[1]
+        step = 1 + n
+        if wrap:
+            end = n * m
+        else:
+            end = n * min(m, n)
+    else:
+        shape = np.array(a.shape)
+
+        if not np.all(np.diff(shape) == 0):
+            raise ValueError("All dimensions of input must be of equal length")
+
+        step = 1 + (np.cumprod(shape[:-1])).sum()
+        end = shape.prod()
+
+    return end, step
+
+@register_jitable
+def _fill_diagonal_scalar(a, val, wrap):
+    end, step = _fill_diagonal_params(a, wrap)
+
+    for i in range(0, end, step):
+        a.flat[i] = val
+
+@register_jitable
+def _fill_diagonal(a, val, wrap):
+    end, step = _fill_diagonal_params(a, wrap)
+    ctr = 0
+    v_len = len(val)
+
+    for i in range(0, end, step):
+        a.flat[i] = val[ctr]
+        ctr += 1
+        ctr = ctr % v_len
+
+@register_jitable
+def _check_val_int(a, val):
+    iinfo = np.iinfo(a.dtype)
+    v_min = iinfo.min
+    v_max = iinfo.max
+
+    # check finite values are within bounds
+    if np.any(~np.isfinite(val)) or np.any(val < v_min) or np.any(val > v_max):
+        raise ValueError('Unable to safely conform val to a.dtype')
+
+@register_jitable
+def _check_val_float(a, val):
+    finfo = np.finfo(a.dtype)
+    v_min = finfo.min
+    v_max = finfo.max
+
+    # check finite values are within bounds
+    finite_vals = val[np.isfinite(val)]
+    if np.any(finite_vals < v_min) or np.any(finite_vals > v_max):
+        raise ValueError('Unable to safely conform val to a.dtype')
+
+# no check performed, needed for pathway where no check is required
+_check_nop = register_jitable(lambda x, y: x)
+
+def _asarray(x):
+    pass
+
+@overload(_asarray)
+def _asarray_impl(x):
+    if isinstance(x, types.Array):
+        return lambda x: x
+    elif isinstance(x, (types.Sequence, types.Tuple)):
+        return lambda x: np.array(x)
+    elif isinstance(x, (types.Float, types.Integer, types.Boolean)):
+        ty = as_dtype(x)
+        return lambda x: np.array([x], dtype=ty)
+
+@overload(np.fill_diagonal)
+def np_fill_diagonal(a, val, wrap=False):
+
+    if a.ndim > 1:
```

Comment:
Please could the `else` for this branch be added to raise a `TypingError` explicitly stating that the first argument must be at least 2-D, thanks.

Without this, the default message will appear:
```
Invalid usage of Function(<function fill_diagonal at 0x7fc72c7cd8c8>) with parameters (array(float64, 1d, C), int64)
 * parameterized
In definition 0:
    All templates rejected
[1] During: resolving callee type: Function(<function fill_diagonal at 0x7fc72c7cd8c8>)
[2] During: typing of call at test_3145.py (6)


File "test_3145.py", line 6:
def foo(x):
    np.fill_diagonal(x, 7)
```
which is fine, but we can say something specific so we probably should.

Reply:
Oh, I see what you mean, I think - so the message now contains: 

```
TypeError: Array must be at least 2-d
```
For consistency with NumPy, I could update this to:

```
ValueError: array must be at least 2-d
```
....if preffered.

Reply:
Thanks for considering this change. The challenge here is to provide feedback whilst in the typing region of an `@overload`. In this part of the `@overload` only the types of the arguments are known and tracing back their actual names is prohibitively difficult from this location. As a result something like:
```python
    else:
        msg = "The first argument must be at least 2-D (found %s-D)" % a.ndim
        raise TypingError(msg)
```
tells the user that it is the first argument that is the problem (as `val` could also be an array?), explains what the problem is and what was presented. In the full context of a traceback it'll look something like:
```python
Invalid usage of Function(<function fill_diagonal at 0x7f15b04e08c8>) with parameters (array(float64, 1d, C), int64)
 * parameterized
In definition 0:
    TypingError: The first argument must be at least 2-D (found 1-D)
    raised from <path>/numba/numba/targets/arraymath.py:1023
[1] During: resolving callee type: Function(<function fill_diagonal at 0x7f15b04e08c8>)
[2] During: typing of call at test_3145.py (6)


File "test_3145.py", line 6:
def foo(x):
    np.fill_diagonal(x, 7)
```
what do you think?

Reply:
Yep, that makes a lot of sense